test/SILGen/capture_order.swift

```diff
@@ -320,6 +312,135 @@ func forward_declared_let_captures() {
+  }
+
+  do {
+    func badFn() -> Any {
+      // expected-error@-1 {{closure captures 'bad' before it is declared}}
+      E.static_gen_fwd { bad }
+      // expected-note@-1 {{captured here}}
+    }
+    let bad = badFn()
+    // expected-note@-1 {{captured value declared here}}
+  }
+
+  do {
+    func badFn() -> Any {
+      // expected-error@-1 {{closure captures 'badNested' before it is declared}}
+      global_fwd { { [badNested] in badNested }() }
+      // expected-note@-1 {{captured here}}
+    }
+    let badNested = badFn()
+    // expected-note@-1 {{captured value declared here}}
+  }
+
+  do {
+    func badFn() -> Any? {
+      // expected-error@-1 {{closure captures 'badOpt' before it is declared}}
+      { () -> Any? in badOpt }()
+      // expected-note@-1 {{captured here}}
+    }
+    let badOpt = badFn()
+    // expected-note@-1 {{captured value declared here}}
+  }
+
+  do {
+    func badFn() -> (Any, Any) {
+      // expected-error@-1 {{closure captures 'badTup' before it is declared}}
+      { (badTup.0, badTup.1) }()
+      // expected-note@-1 {{captured here}}
+    }
+    let badTup = badFn()
+    // expected-note@-1 {{captured value declared here}}
+  }
+
+  do {
+    func badFn() -> (Int, Any) {
+      // expected-error@-1 {{closure captures 'badTup' before it is declared}}
+      { (badTup.0, badTup.1) }()
+      // expected-note@-1 {{captured here}}
+    }
+    let badTup = badFn()
+    // expected-note@-1 {{captured value declared here}}
+  }
+
+  do {
+    func badFn() -> (Any, Any) {
+      // expected-error@-1 {{closure captures 'badTup3' before it is declared}}
+      // expected-error@-2 {{closure captures 'badTup4' before it is declared}}
+      { (badTup4, badTup3) }()
+      // expected-note@-1 2{{captured here}}
+    }
+    let (badTup3, badTup4) = badFn()
+    // expected-note@-1 2{{captured value declared here}}
+  }
+
+  do {
+    struct S { var p: Any }
+    func badFn() -> S {
+      // expected-error@-1 {{closure captures 'badStruct' before it is declared}}
+      { S(p: badStruct.p) }()
+      // expected-note@-1 {{captured here}}
+    }
+    let badStruct = badFn()
+    // expected-note@-1 {{captured value declared here}}
+  }
+
+  do {
+    enum EE {
+      case boring
+      case weird(Any)
+      case strange(Any)
+    }
+
+    func badFn() -> EE {
+      // expected-error@-1 {{closure captures 'badEnum' before it is declared}}
+      { .weird(EE.strange(badEnum)) }()
+      // expected-note@-1 {{captured here}}
+    }
+    let badEnum = badFn()
+    // expected-note@-1 {{captured value declared here}}
+  }
+
+  do {
+    func badFn() -> any P {
+      // expected-error@-1 {{closure captures 'badproto' before it is declared}}
+      global_fwd_p { badproto }
+      // expected-note@-1 {{captured here}}
+    }
+    let badproto = badFn()
+    // expected-note@-1 {{captured value declared here}}
+  }
+}
+
 // FIXME: should/could these be diagnosed instead of accepted?
```

Comment:
tangential: after thinking about these cases a bit more, i'm not sure the commentary is warranted (i.e. recording my past confusion), and perhaps they should be removed. given how the implementation of lazy variables works, i don't think these have the same issues as the non-lazy cases. the first one will infinitely recurse, but that's 'fine', and the second one is confusing, but there's a level of indirection where the actual storage is the implicitly-synthesized lazy storage variable.

Reply:
Yeah I would love if we could ban forward referencing local vars across the board in Sema